src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -24,8 +29,45 @@ public synchronized Product getProduct(Long id) throws NoSuchElementException {
     }
 
     @Override
-    public synchronized List<Product> findProducts() {
-        return products;
+    public synchronized List<Product> findProducts(String query, SortField sortField, SortOrder sortOrder) {
+        HashMap<Product, Integer> productMap = new HashMap<>();
+        if (query != null) {
+            for (int i = 0; i < products.size(); i++) {
+                for (int j = 0; j < query.split(" ").length; j++) {
+                    if (products.get(i).getDescription().toUpperCase().contains((query.split(" ")[j]).toUpperCase())) {
+                        if (products.get(i).getDescription().toUpperCase().equals(query.toUpperCase())) {
+                            productMap.put(products.get(i), Integer.MAX_VALUE); // костыль, если запрос полностью совпадает названию продукта то он отобразиться первым
+                        } else if (productMap.containsKey(products.get(i))) {
+                            productMap.put(products.get(i), productMap.get(products.get(i)) + 1); //иначе добавить продукт в map либо увеличить число совпадений для продукта на 1
+                        } else productMap.put(products.get(i), 0);
+                    }
+                }
+            }
+        }
```

Comment:
1) Can we write this using streams? If you are not sure, ask me for details how this can be done.

2) for (int j = 0; j < query.split(" ").length; j++) - we are calling this EACH for each product - this affects performance.

3) Repeating calls: 
products.get(i).getDescription().toUpperCase()
query.toUpperCase()
products.get(i)

We really don't need to call each of these n*m times (for i from 0 to n; for j from 0 to M), right?
Put them into variables and reuse.